modules/authglobus/lib/Auth/Source/Globus.php

```diff
@@ -0,0 +1,213 @@
+<?php
+
+require_once(dirname(dirname(dirname(dirname(dirname(__FILE__))))) . '/oauth/lib/Consumer.php');
+require_once(dirname(dirname(dirname(dirname(dirname(__FILE__))))) . '/oauth/libextinc/OAuth.php');
+
+/**
+ * Authenticate using Globus Auth
+ * Globus Auth Documentation: https://docs.globus.org/api/auth/developer-guide
+ *
+ * @author Rudra Chakraborty, University at Buffalo, Center for Computational Research. rudracha@buffalo.edu
+ * @package SimpleSAMLphp
+ */
+
+class sspmod_authglobus_Auth_Source_Globus extends SimpleSAML_Auth_Source
+{
+    /**
+     * The string used to identify our states.
+    */
+    const STAGE_INIT = 'Globus:init';
+
+    /**
+     * The key of the AuthId field in the state.
+    */
+    const AUTHID = 'Globus:AuthId';
+
+    /**
+     * Globus Endpoint
+    */
+    const API_ENDPT = 'https://auth.globus.org/v2';
+
+    private $key;
+    private $secret;
+    private $scope;
+    private $reponse_type;
+    private $redirect_uri;
+
+    /**
+     * Constructor for this authentication source.
+     *
+     * @param array $info  Information about this authentication source.
+     * @param array $config  Configuration.
+     */
+    public function __construct($info, $config) {
+        // Call the parent constructor first, as required by the interface
+        parent::__construct($info, $config);
+        
+        $this->key = $config['key'];
+        $this->secret = $config['secret'];
+        $this->scope = $config['scope'];
+        $this->response_type = $config['response_type'];
+        $this->redirect_uri= $config['redirect_uri'];
+    }
+
+    /**
+     * Flattens an associative array, derived from LinkedIn module's flatten
+     *
+     * @param array $array
+     * @param string $prefix
+     *
+     * @return array the array with the new concatenated keys
+     */
+    protected function flatten($array, $prefix = '')
+    {
+        $newArr = array();
+        foreach ($array as $key => $value) {
+            if (is_array($value)) {
+                $newArr = $newArr + $this->flatten($value, $prefix . $key . '.');
+            } else {
+                $newArr[$prefix . $key] = $value;
+            }
+        }
+        return $newArr;
+    }
+
+    /**
+     * Exchange an authorization code for Globus Tokens.
+     *
+     * @param array $contents The contents of the POST request. REF: https://docs.globus.org/api/auth/reference/#authorization_code_grant_preferred
+     * @param string $signature Your Client ID + Secret from the Globus app you've registered.
+     */
+    private function getGlobusTokens($contents, $signature) {
+        $curl = curl_init();
```

Comment:
we should probably make this generic so that curl can be used in multiple places.